File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/membership_plan.dart
import 'package:equatable/equatable.dart';

class MembershipPlan extends Equatable {
  final int id;
  final int outletId;
  final String name;
  final double price;
  final int durationDays;
  final bool isActive;
  final double discountPercentage;
  final String? description;
  final int? level;
  final String? createdAt;
  final String? updatedAt;
  final int membershipContractsCount;

  const MembershipPlan({
    required this.id,
    required this.outletId,
    required this.name,
    required this.price,
    required this.durationDays,
    required this.isActive,
    required this.discountPercentage,
    this.description,
    this.level,
    this.createdAt,
    this.updatedAt,
    this.membershipContractsCount = 0,
  });

  factory MembershipPlan.fromModel(dynamic model) {
    return MembershipPlan(
      id: model.id,
      outletId: model.outletId,
      name: model.name,
      price: model.price,
      durationDays: model.durationDays,
      isActive: model.isActive,
      discountPercentage: model.discountPercentage,
      description: model.description,
      level: model.level,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      membershipContractsCount: model.membershipContractsCount,
    );
  }

  @override
  List<Object?> get props => [
    id,
    outletId,
    name,
    price,
    durationDays,
    isActive,
    discountPercentage,
    description,
    level,
    createdAt,
    updatedAt,
    membershipContractsCount,
  ];
}

File: laundry/wash_wallet_be/resources/js/types/membership_plan.ts
import { BaseFilters, BaseSortOptions, MembershipContract, Outlet } from ".";

/**
 * MembershipPlan interface
 */
export interface MembershipPlan {
    id: number;
    outletId: number;
    name: string;
    price: number;
    durationDays: number;
    isActive: boolean;
    discountPercentage: number;
    level: number;
    description?: string | null;
    createdAt: string;
    updatedAt: string;
    outlet: Outlet;
    membershipContracts?: MembershipContract[];
    membershipContractsCount?: number;
    activeMembershipContractsCount?: number;
}

/**
 * MembershipPlan filter interface extends complete base filters
 */
export interface MembershipPlanFilters extends BaseFilters {
    outletId?: number;
    minPrice?: number;
    maxPrice?: number;
    isActive?: boolean;
    minDurationDays?: number;
    maxDurationDays?: number;
    minDiscountPercentage?: number;
    maxDiscountPercentage?: number;
}

/**
 * MembershipPlan sort options
 */
export interface MembershipPlanSortOptions extends BaseSortOptions {
    column:
        | "name"
        | "price"
        | "durationDays"
        | "discountPercentage"
        | "createdAt"
        | "updatedAt"
        | "isActive"
        | "popular";
}

/**
 * MembershipPlan form data interface
 */
export interface MembershipPlanFormData {
    outletId: number;
    name: string;
    price: number;
    durationDays?: number | null;
    isActive?: boolean | null;
    discountPercentage: number;
    description?: string;
    level: number;
}
